internal/pkg/policy/policy_output.go

```diff
@@ -137,6 +192,104 @@ func (p *PolicyOutput) Prepare(ctx context.Context, zlog zerolog.Logger, bulker
 	return nil
 }
 
+func fetchAPIKeyRoles(ctx context.Context, b bulk.Bulk, apiKeyID string) (*RoleT, error) {
+	res, err := b.APIKeyRead(ctx, apiKeyID, true)
+	if err != nil {
+		return nil, err
+	}
+
+	roleMap, err := smap.Parse(res.RoleDescriptors)
+	if err != nil {
+		return nil, err
+	}
+	r := &RoleT{
+		Raw: res.RoleDescriptors,
+	}
+
+	// Stable hash on permissions payload
+	if r.Sha2, err = roleMap.Hash(); err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
+func mergeRoles(zlog zerolog.Logger, old, new *RoleT) (*RoleT, error) {
+	if old == nil {
+		return new, nil
+	}
+	if new == nil {
+		return old, nil
+	}
+
+	oldMap, err := smap.Parse(old.Raw)
+	if err != nil {
+		return nil, err
+	}
+	if oldMap == nil {
+		return new, nil
+	}
+
+	newMap, err := smap.Parse(new.Raw)
+	if err != nil {
+		return nil, err
+	}
+
+	if newMap == nil {
+		return old, nil
+	}
+
+	destMap := smap.Map{}
+	// copy all from new
+	for k, v := range newMap {
+		destMap[k] = v
+	}
+
+	findNewKey := func(m smap.Map, candidate string) string {
+		// creates new key for permissions
+		// marks it stale so they can be removed later
+		if _, exists := m[candidate]; !exists {
+			return fmt.Sprintf("%s-rdstale", candidate)
+		}
+
+		// 1 should be enough, 100 is just to have some space
+		for i := 0; i < 100; i++ {
+			c := fmt.Sprintf("%s-%d-rdstale", candidate, i)
+
+			if _, exists := m[c]; !exists {
+				return c
+			}
+		}
+
+		return ""
+	}
+	// copy old
+	for k, v := range oldMap {
+		newKey := findNewKey(destMap, k)
+		if newKey == "" {
+			zlog.Warn().Msg("Failed to find a key for role assignement.")
+
+			zlog.Debug().
+				RawJSON("roles", new.Raw).
+				Str("candidate", k).
+				Msg("Failed to find a key for role assignement.")
+
+			continue
+		}
+		destMap[newKey] = v
+	}
```

Comment:
I didn't understand how it works

Reply:
added comments describing merge strategy